### PR TITLE
Optimizing AttributedSpans.collapseSpans() to run in linear time

### PR DESCRIPTION
**Pull request by jmatth**
The current implementation of `collapseSpans` operates by pre-generating the list of start and end points for the collapsed spans, then walking that list and repeatedly checking the list of `SpanMarker`s to determine which `Attribution`s should apply to each span. In cases with many spans this causes a large amount of duplicate work, especially in cases with overlapping spans. In my testing it takes the method almost 2 seconds to return with 1000 spans. For a less contrived example, it only needs about 50 overlapping spans to take over 16ms (time for a frame when running at 60fps).

This PR:

- Changes `collapseSpans` to run with a single linear pass of the `SpanMarker` list. It also eliminates the preprocessing step of generating the start and end locations for the collapsed spans, which wasn't significantly contributing to the runtime but fewer LoC is always a nice bonus.
- Updates the sorting logic for `SpanMarker` to always have `start` markers precede `end` markers. Without this the inner loop of `collapseSpans` needs some ugly conditionals to handle an edge case where an end marker precedes a start marker with the same offset.
- Adds a couple tests to the AttributedSpans test suite for behavior that that wasn't already covered but broke tests from other parts of the library while I was working on this.
- Adds some minor additions to `MultiAttributionSpan`: `copyWith` for convenient copying, `toString` override for better log messages.

After these changes it the method takes 3ms to return with 50 spans and 30ms with 1000. I ran the tests locally via `flutter test` and everything passed.

### CODE REVIEW

File: super_editor/lib/src/infrastructure/attributed_spans.dart [hunk 1]
```diff
@@ -820,89 +818,60 @@ class AttributedSpans {
...
-      _log.fine('end points after change: $endPoints');
-    }
-    if (!endPoints.contains(contentLength - 1)) {
-      // This condition occurs when there are no attributions.
-      _log.fine('adding a final endpoint at end of text');
-      endPoints.add(contentLength - 1);
+      // Because we handle committing completed spans before this, we know that by this point the current marker should
+      // indicate changes to the current span.
+      if (marker.isStart) {
+        // Merge the new attribution into the current span.
```
Comment (matthew-carroll) [inline super_editor/lib/src/infrastructure/attributed_spans.dart:858]: Can you change the word "merge" in this comment and the log statement to "add" so that is clearly opposes "remove" in the else statement?

File: super_editor/lib/src/infrastructure/attributed_spans.dart [hunk 2]
```diff
@@ -820,89 +818,60 @@ class AttributedSpans {
...
       }
 
-      _log.fine('start points after change: $startPoints');
-      _log.fine('end points after change: $endPoints');
-    }
-    if (!endPoints.contains(contentLength - 1)) {
-      // This condition occurs when there are no attributions.
-      _log.fine('adding a final endpoint at end of text');
-      endPoints.add(contentLength - 1);
+      // Because we handle committing completed spans before this, we know that by this point the current marker should
```
Comment (matthew-carroll) [inline super_editor/lib/src/infrastructure/attributed_spans.dart:855]: This comment is probably important, but I'm not entirely sure what it's trying to say. Can you revisit this comment and try to establish some context? The comment seems to assume that the reader already has a problem in mind, but I'm not sure what that problem is.

File: super_editor/lib/src/infrastructure/attributed_spans.dart [hunk 3]
```diff
@@ -820,89 +818,60 @@ class AttributedSpans {
...
+      }
     }
 
-    if (startPoints.length != endPoints.length) {
-      _log.fine('start points: $startPoints');
-      _log.fine('end points: $endPoints');
-      throw Exception(
-          ' - mismatch between number of start points and end points. Content length: $contentLength, Start: ${startPoints.length} -> $startPoints, End: ${endPoints.length} -> $endPoints, from attributions: $_attributions');
+    if (collapsedSpans.isEmpty || collapsedSpans.last.end < contentLength - 1) {
+      // The spans committed during the loop didn't cover the entire range.  The value in currentSpan should already
```
Comment (matthew-carroll) [inline super_editor/lib/src/infrastructure/attributed_spans.dart:869]: The language in this comment mixes some times and responsibilities, making it difficult to differentiate between the description of the current situation vs the description of what needs to be done. Can you revisit this comment and try to make it clear what the situation is when coming into the if-statement, vs the work that you're about to do?

File: super_editor/lib/src/infrastructure/attributed_spans.dart [hunk 4]
```diff
@@ -820,89 +818,60 @@ class AttributedSpans {
...
+        _log.fine('removing attribution ${marker.attribution}, current span is now $currentSpan.');
+      }
     }
 
-    if (startPoints.length != endPoints.length) {
-      _log.fine('start points: $startPoints');
-      _log.fine('end points: $endPoints');
-      throw Exception(
-          ' - mismatch between number of start points and end points. Content length: $contentLength, Start: ${startPoints.length} -> $startPoints, End: ${endPoints.length} -> $endPoints, from attributions: $_attributions');
+    if (collapsedSpans.isEmpty || collapsedSpans.last.end < contentLength - 1) {
```
Comment (matthew-carroll) [inline super_editor/lib/src/infrastructure/attributed_spans.dart:868]: Is it possible to check for the empty situation near the beginning of the method, similar to content length? Or is that calculation too difficult to do on its own?

**Review by matthew-carroll — CHANGES_REQUESTED** [review]
Thanks for the improvements. I added a few requests for minor changes.

File: super_editor/lib/src/infrastructure/attributed_spans.dart [hunk 5]
```diff
@@ -820,89 +818,60 @@ class AttributedSpans {
...
-          endPoints.add(marker.offset - 1);
-        }
-      }
-      if (marker.isEnd) {
-        // Add an `end` point.
-        if (!endPoints.contains(marker.offset)) {
-          _log.fine('adding an end point at: ${marker.offset}');
-          endPoints.add(marker.offset);
-        }
+        // If we encountered a start token, end the current span 1 index earlier to simulate the missing end marker.
```
[the remark this answers is not in the document]
Comment (jmatth) [reply super_editor/lib/src/infrastructure/attributed_spans.dart:840]: Yeah, this is probably a quirk of the development process working its way into the comments. In an early implementation I was preprocessing the list to insert start and end markers at the points where spans overlapped. I got rid of that because 1) it's actually unnecessary and 2) it's actually harder to read and reason about. Since those explicit start/end markers aren't there, this on the fly calculation is now "simulating" them.

I agree this wording doesn't make sense now without that particular thought process. I'll update it similar to your suggestion on the end token.

File: super_editor/lib/src/infrastructure/attributed_spans.dart (hunk 2 above)
Comment (jmatth) [reply super_editor/lib/src/infrastructure/attributed_spans.dart:855]: The comment is there to say that because the way the code is ordered, by the time we reach this point we know that we can just apply changes to the current span based on the current marker without any other checks. A previous version of the code required a bounds check on the current marker's offset, and this comment came about when that check was made unnecessary and removed.

How about: `By the time we get here, we are guaranteed that the current marker should modify the current span. Apply changes based on the type of the marker.`?

File: super_editor/lib/src/infrastructure/attributed_spans.dart [hunk 6]
```diff
@@ -820,89 +818,60 @@ class AttributedSpans {
...
-        // `end` marker.
-        if (marker.offset < contentLength - 1 && !startPoints.contains(marker.offset + 1)) {
-          _log.fine('jumping forward one to add a start point at: ${marker.offset + 1}');
-          startPoints.add(marker.offset + 1);
-        }
+        // Commit the completed span.
+        collapsedSpans.add(currentSpan.copyWith(end: currentEnd));
+        _log.fine('committed span ${collapsedSpans.last}');
+
+        // If we encountered an end token, start the next span 1 index later to simulate the missing start marker.
```
[the remark this answers is not in the document]
Comment (jmatth) [reply super_editor/lib/src/infrastructure/attributed_spans.dart]: Addressed in [this comment](https://github.com/superlistapp/super_editor/pull/407#discussion_r798726573).

File: super_editor/lib/src/infrastructure/attributed_spans.dart (hunk 2 above)
Comment (matthew-carroll) [reply super_editor/lib/src/infrastructure/attributed_spans.dart]: Sure, that comment makes sense

File: super_editor/lib/src/infrastructure/attributed_spans.dart (hunk 4 above)
Comment (matthew-carroll) [reply super_editor/lib/src/infrastructure/attributed_spans.dart]: I think I'd prefer this approach. This way readers know that if they get past the guard statements, they're definitely doing some span calculations.

Comment (jmatth) [reply super_editor/lib/src/infrastructure/attributed_spans.dart]: Cool, taken care of in b3c9c1ee02162dd599d82821a568c042a7490af6.

File: super_editor/lib/src/infrastructure/attributed_spans.dart (hunk 2 above)
Comment (jmatth) [reply super_editor/lib/src/infrastructure/attributed_spans.dart]: Taken care of in f7a40f556656ae15bb4964cbb1fe7afb2f8cc931.

**Comment by jmatth** [comment]
I think I've addressed all the outstanding PR comments.

**Review by matthew-carroll — APPROVED** [review]
LGTM. Thanks.
